### DISCUSSION TITLE
Modbus RTU for the noob

### DISCUSSION THREAD

**Discussion by varidel** [eModbus/eModbus#152] [Q&A]
hey i'am a French student and I have a project for my graduate studies where I have to realize an Rs485 line in modbus RTU 
The problem is that I have never done object programming and have a hard time understanding some of the code in the examples. Can anyone help me? 
Thanks

**Comment by Miq1** [comment]
Well, why not? Just give us some more details about your project. ESP32, I suppose? Which RS485 adaptors? Do you have a data sheet on the Modbus servers you are going to use?

**Reply by varidel** [reply]
hello i.  send my mail for easier contact mehdisaghi@hotmail.fr

**Reply by varidel** [reply]
I am using an esp 32 in RS485 with an adapter that I have to create consisting of a SP3481 and I don't have a datasheet for modbus

**Reply by Miq1** [reply]
The SP3481 datasheet can easily be found on the net. I had a quick look at it: the chip looks straightforward enough to wire. Like any other RS485 adapter, the DI and RO lines will have to be connected to one of the ESP32's UARTs TX and RX pins, the DE and RE lines may be jointly connected to one GPIO as toggle pin. On the RS485 side the A/B lines need to be properly terminated by an 120 Ohms resistor between both.
What devices will you have connected to the RS485 bus? I suppose the ESP32 is meant to be the client on the bus?

**Reply by varidel** [reply]
I'm sorry, I've already done my research on the Sp3485 and I've come to the same conclusion as you, but my problem is more with the coding. To put it simply, I have a connected spa and I need to store the values of the temperature, the values of the leds and of the bp's, and as soon as they are requested, I have to supply them.

**Reply by varidel** [reply]
I think I'll use the emodbus library but I don't understand this and I'm not sure but I don't think it's the only library to use

**Reply by Miq1** [reply]
> I'm sorry, I've already done my research on the Sp3485 and I've come to the same conclusion as you, but my problem is more with the coding. To put it simply, I have a connected spa and I need to store the values of the temperature, the values of the leds and of the bp's, and as soon as they are requested, I have to supply them.

A spa? You mean, a whirlpool or the like? What are BPs?

The final sentence you wrote more looks like you are to build a Modbus server, that will provide the values upon Modbus request, but not a client that is requesting values. Is that correct?

So basically you will need to write a part that will collect the values from the spa and keep these in memory, while the Modbus server will provide the data for responses. Since you did not tell anything yet on registers, function codes or server IDs required, we cannot model more at the moment. Would you be able to give us the text of the task you are given?

**Reply by varidel** [reply]
the BP is a push button and it is a whirlpool

you evoke one of my problem at the level of the registers, I do not understand how to create registers, the codes functions I made some research but without forcefully to understand completely concern the IDs I am not yet searched  


The explanations in my contract file are quite vague as nothing is said specifically except that I have to develop the man-machine interface in RS485

**Reply by Miq1** [reply]
It would be helpful, though, to see what really is requested...

Anyway, some basics:
- Modbus data is transmitted in 16 bit "registers" at "addresses" - a pure virtual definition. It is completely up to you to define at what address (aka "register number") you will provide whatever value you chose. The addresses or register boundaries have no impact at all on the layout of the memory you are going to use in your program. 
The example server sketches simply have a vector of values in memory that is read out in the order of register addresses. This is one common concept to implement that.
- Data not fitting into 16 bit lengths may be spread across several "registers". Again it is you defining what registers will be used for what data.
- Same applies to function codes. You decide which function codes are to be honored. Commonly used to request data are 0x03 READ_HOLD_REGISTER and 0x04 READ_INPUT_REGISTER. The library will respond to any function code not allowed by you with an appropriate error response.
- Server IDs need to be unique on a single RTU Modbus, so it may be advisable to be able to have that changed at runtime, to not collide with another server on the same bus using that very ID. IDs may range from 1 to 247.
